buildchain/buildchain/deps.py

```diff
@@ -0,0 +1,36 @@
+from doit.exceptions import TaskError
+from pathlib import Path
+from typing import Optional
+import shutil
+from buildchain import config
+
+
+def task_check_for():
+
+    def get_command_location(command_name: str, binary_path: Optional[str]
+                             ) -> Optional[TaskError]:
+        """Search the location of the command `command_name`.
+
+        If the user already specified a path for it, check for the existence
+        of the binary at the given location. Otherwise, look it up from PATH.
+
+        If the binary is found, the `CMD` dict is updated, otherwise it fails.
```

Comment:
I'm quite lost. Let me detail the few code paths I understand of your code:

_Case 1:_ `SOME_CMD_BIN` is not defined in the environment, so this task will search in `$PATH` and override `CMD['some-cmd']`

_Case 2_: `SOME_CMD_BIN` is set to `/some/path/some-cmd`, so this task will search in `/some/path` and set `CMD['some-cmd']` to the same value as it was already

_Case 3_: `SOME_CMD_BIN` is set to `/some/path/alias-cmd` (note the different binary name), so this task will search for `some-cmd` in `/some/path`, and **will fail**.

I believe we need to be more declarative and less magic about finding binaries we rely on.

First, using a mutable dict like this prevents static checkers from knowing whether the relevant command is defined in `config.CMD`. We could have used an `enum.Enum` to fix that aspect.

Second, you are mutating the dict at every execution of `check_for:some-cmd`, redundant though not critical.

Finally, I'd rather let the user responsible of managing his commands: the task could just _check_ if the provided path exists, complain if not, and get on with it. No need to mutate anything, nor to run this check for each task that rely on it.

Reply:
@gdemonet I would certainly look into the potential use of enumerators.

In regards to the second potential issue of mutating Dicts each time; I think this is pinned to the underline functioning of **Python Doit**.

**step i.** All tasks are created + graphs are built.
**step ii.** Tasks are executed.

What does this mean for us??
Once tasks are created, the dict is initialized with none until the tasks are ran  For example, the dict will only be updated if  `check_for:docker` had ran before hence the mutation:

`TaskError - taskid:check_for:skopeo
command skopeo {'docker': '/usr/bin/docker', 'skopeo': None, 'vagrant': None, 'mkisofs': None, 'hardlink': None} not found in $PATH
`

@slaperche-scality has proposed a better way of handling this and we are looking into this.

And yes with this new approach, we potentially will take off mutations each time using `        'actions': [CmdAction(create_cmd_string)],`

Reply:
> Case 2: SOME_CMD_BIN is set to /some/path/some-cmd, so this task will search in /some/path and set CMD['some-cmd'] to the same value as it was already

Yeah, we don't change the value here.
The point is to check that we indeed have a `some-cmd` and that it's executable, … in the given path.

> No need to mutate anything, nor to run this check for each task that rely on it.
You need to run it for every task that rely on it because you don't know WHICH command will be run, and as said before it should be run only once per execution.

> Second, you are mutating the dict at every execution of check_for:some-cmd, redundant though not critical.

For that point, we could probably have a custom `uptodate` that run the task only if the dict entry is `None`: that way it'll only be run once, even if multiple tasks depends on it.
But I think even as it is, the task is only run once (to be confirmed) so it's unnecessary.

Reply:
> In regards to the second potential issue of mutating Dicts each time; I think this is pinned to the underline functioning of Python Doit.
>
> **step i.** All tasks are created + graphs are built.
> **step ii.** Tasks are executed.

I don't see how this means you **need** mutability. Most importantly, my point is that we don't need to mutate _anything_.

> For that point, we could probably have a custom `uptodate` that run the task only if the dict entry is `None`: that way it'll only be run once, even if multiple tasks depends on it.
> But I think even as it is, the task is only run once (to be confirmed) so it's unnecessary.

If the task is run once only, then let it fail if the path is not found, and that's enough. No need to store it in a dict (either it's in `$PATH`, or it's provided by `os.getenv("XXX_BIN")`).

Reply:
Also, how would this behaviour of mutating dicts work in a concurrent execution? Is `doit` responsible for passing the shared state to spawned tasks?

Reply:
> If the task is run once only, then let it fail if the path is not found, and that's enough. No need to store it in a dict (either it's in $PATH, or it's provided by os.getenv("XXX_BIN")).

Hum, fair enough.
This will simplify the code!

> Also, how would this behaviour of mutating dicts work in a concurrent execution? Is doit responsible for passing the shared state to spawned tasks?

Good point.